engine/access/state_stream/subscription.go

```diff
@@ -0,0 +1,115 @@
+package state_stream
+
+import (
+	"context"
+	"time"
+
+	"github.com/google/uuid"
+)
+
+// GetDataByHeightFunc is a callback used by subscriptions to retrieve data for a given height.
+// Expected errors:
+// - storage.ErrNotFound
+// - execution_data.BlobNotFoundError
+// All other errors are considered exceptions
+type GetDataByHeightFunc func(ctx context.Context, height uint64) (interface{}, error)
+
+// Subscription represents a streaming request, and handles the communication between the grpc handler
+// and the backend implementation.
+type Subscription interface {
+	// ID returns the unique identifier for this subscription used for logging
+	ID() string
+
+	// Channel returns the channel from which subscriptino data can be read
+	Channel() <-chan interface{}
+
+	// Err returns the error that caused the subscription to fail
+	Err() error
+}
+
+type SubscriptionImpl struct {
+	id string
+
+	// ch is the channel used to pass data to the receiver
+	ch chan interface{}
+
+	// err is the error that caused the subscription to fail
+	err error
+}
+
+func NewSubscription() *SubscriptionImpl {
```

Comment:
nit: this seems like a good usecase for generics: `func NewSubscriptionFor[T any]() *SubscriptionImpl[T]`